client/state/push-notifications/schema.js

```diff
@@ -1,13 +1,10 @@
 const boolType = { type: 'boolean' };
 const stringType = { type: 'string' };
-const numberType = { type: 'number' };
 
 export const settingsSchema = {
 	type: 'object',
 	properties: {
 		enabled: boolType,
-		dismissedNotice: boolType,
-		dismissedNoticeAt: numberType,
 		showingUnblockInstructions: boolType,
 	},
 	additionalProperties: false,
```

Comment:
If additional properties are forbidden here, then the reducer will fail to load the persisted state that was stored before this PR was merged, is that right? There are `dismissedNotice` fields in the persisted data, but that won't pass the schema validation now.

Reply:
Yes, that's right. We're removing those fields because we don't need them anymore. And like with any other schema changes, this will invalidate any persisted data, which I'm totally fine with.

Reply:
My point though is that the persisted data are not invalid, we just stopped using some fields, yes, but the remaining ones are perfectly OK. Change doesn't necessarily mean invalidation.

Reply:
I see. So you would recommend keeping the old schema to prevent the invalidation?

Reply:
The reason why I think it's not a good idea is that removing `additionalProperties: false` will basically remove half of the utility of the schema we have in place. On top of the persistence, the schema validation is currently responsible for:

* Making sure data contains the necessary fields.
* Making sure data doesn't contain any unnecessary fields.

And with your suggestion, we're giving up the second bit. 

Instead, I'm happy to invalidate the persisted data once but to keep it cleaner and smaller for any subsequent operations.

What are your thoughts?

Reply:
> Making sure data doesn't contain any unnecessary fields.

But extra fields don't make the data invalid. There is, e.g., a `dismissedNotice` field, we no longer use it, but the `enabled` value is still OK, we don't need to drop it. Once the new Calypso writes the state back to storage, the removed fields will no longer be saved, but the `enabled` value will survive.

Reply:
I disagree here. We change the schema to adapt to the fields we actually need and use, so the old unnecessary data should be deleted. If we keep those extra unnecessary fields, why do we even have a schema? Why don't we just persist all the data that comes in and out, regardless of its format? Also, I don't think it causes such a substantial pain to drop the existing persisted data in favor of fresh data.

Reply:
I think I'm starting to understand your point of view 🙂 In the `system` and `settings` reducers, there is not a subreducer for each field. So, once we load a legacy `dismissedNotice` field, it will stay there forever, because it's a property of the state object.

On the other hand, if there was a `dismissedNotice` sub-reducer, its removal would mean that `dismissedNotice` is no longer in the state and is never saved.

Ideally, we should be able to load old persistent state with legacy fields, and then save a new state without these fields. But that's not easy to do with how the code is currently written.

> If we keep those extra unnecessary fields, why do we even have a schema?

It there to catch incompatible changes. Like when `foo` field changes from `number` to `string`. Then, loading an old persisted state would store a number in `state.foo`. But it's supposed to be `string`. It's no longer safe to call, e.g., `state.foo.includes( 'bar' )`. If we validate `foo` with a schema, assuming that `state.foo` is always a string becomes safe again.